File: 0205-isomorphic-strings/0205-isomorphic-strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        unordered_map<char, char> s_to_t;
        unordered_map<char, char> t_to_s;
        
        for (size_t i = 0; i < s.length(); ++i) {
            char char_s = s[i];
            char char_t = t[i];

            if (s_to_t.find(char_s) != s_to_t.end()) {
                if (s_to_t[char_s] != char_t) {
                    return false;
                }
            } else {
                s_to_t[char_s] = char_t;
            }

            if (t_to_s.find(char_t) != t_to_s.end()) {
                if (t_to_s[char_t] != char_s) {
                    return false;
                }
            } else {
                t_to_s[char_t] = char_s;
            }
        }

        return true;
    }
};

File: 0205-isomorphic-strings/0205-isomorphic-strings.rs
use std::collections::HashMap;
impl Solution {
    pub fn is_isomorphic(s: String, t: String) -> bool {
        let mut s_to_t = HashMap::new();
        let mut t_to_s = HashMap::new();

        for (char_s, char_t) in s.chars().zip(t.chars()){
            if let Some(&mapped_char_t) = s_to_t.get(&char_s) {
                if mapped_char_t != char_t{
                    return false;
                }
            } else {
                s_to_t.insert(char_s, char_t);
            }

            if let Some(&mapped_char_s) = t_to_s.get(&char_t) {
                if mapped_char_s != char_s {
                    return false;
                } 
            } else {
                    t_to_s.insert(char_t, char_s);
                }
        }
        true
    }
}